New story-specific params/metadata

kyanmichael opened the pull request:
This PR consists of three categories:


## 1. Renamed internal items

The purpose of this renaming is to improve the codebase's [greppability](http://john.freml.in/grep-orientated-programming):

> Greppability is the ease with which one can navigate a body of source code just by searching simple text keywords. From determining what code caused an output to tracing all the callers of a function, there's plenty that can be possible by text searching - or not depending how names are used or the project structured. And the easier it is, the faster and more reliably new developers can be productive.

For example, the name `setThemes` is currently used for two different things in `src/Themes.tsx`:

1. The function which updates the `themes` state variable.
2. The channel event which is received when the `src/ThemesProvider.tsx` component mounts.

By giving these items unique (and more specific) names, the codebase's greppability is improved.

**Note: Only internal items have been renamed. No items exposed in the addon's public API have been renamed, so this is not a breaking change.**

Commits:

* cdbcd20  –  Rename `themes` prop to `decoratorThemes`
* cf6225f  –  Rename `theme` to `stateTheme`
* 0e56d61  –  Rename `setTheme` to `setStateTheme`
* eb5bcc1  –  Rename `setThemes` to `decoratorThemesReceived`
* 7c49044  –  Rename `selectTheme` to `panelThemeSelected`
* e48f1f5  –  Rename `theme` to `stateTheme`
* fc3746e  –  Rename `setTheme` to `setStateTheme`
* bacde06  –  Rename `themes` to `stateThemes`
* cf2c742  –  Rename `setThemes` to `setStateThemes`

Complete diff – https://github.com/Sage/storybook-addon-styled-component-theme/compare/5678bfb...cf2c742


## 2. Miscellaneous improvements

* 3998235  –  Refactor `fontFamily` to avoid TSLint issue
* 421ffef  –  Improve React keys for theme buttons


## 3. Implemented new story-specific params/metadata

**Note: These new params are all optional, so this is not a breaking change.**

The new params are documented in the updated `README.md`:

> The following fields (all optional) can be supplied in a `themeSelector` metadata object for a story:
>
> * **`themes`**  –  An array of themes, overriding the `themes` passed to `addDecorator(withThemesProvider(themes))` (if any).
> * **`singleThemeMessage`**  –  A `string`, containing a message to be displayed in the "Themes" Panel when there's only one theme available.
> * **`showSingleThemeButton`**  –  A `boolean` (default `true`), specifying whether to display the single theme button when there's only one theme available.
>
> For example:
>
> ```jsx
> const customThemes = [theme3, theme4];
>
> storiesOf("demo", module)
>   .add("demo div", () => <div>DEMO</div>, {
>     themeSelector: {
>       themes: customThemes
>     }
>   });
> ```
>
> ```jsx
> const customThemes = [theme5];
>
> storiesOf("demo", module)
>   .add("demo div", () => <div>DEMO</div>, {
>     themeSelector: {
>       themes:                customThemes,
>       singleThemeMessage:    "There is only one theme configured for this story.",
>       showSingleThemeButton: false
>     }
>   });
> ```

Commits:

* 8d83996  –  `panelThemeSelected`: pass theme instead of name
* 23efc4b  –  Create `src/constants.tsx`
* f84dafa  –  Implement `storyThemesReceived` functionality
* fe9b282  –  `withThemesProvider`: default value for `themes`
* b7de211  –  Implement `singleThemeMessage` functionality
* a7794ed  –  Implement `showSingleThemeButton` functionality
* 20d3b93  –  Update `Themes.spec.tsx`
* 890d82b  –  Update `README.md`
* a3dc4c7  –  `onReceiveThemes()`: need to call `setQueryParams()`
* e4a1df9  –  Minor cleanup in `README.md` and `src/Themes.tsx`

Commit:
(storyspecific-themes) Rename `themes` prop to `decoratorThemes`

Commit:
(storyspecific-themes) Rename `theme` to `stateTheme`

Commit:
(storyspecific-themes) Rename `setTheme` to `setStateTheme`

Commit:
(storyspecific-themes) Rename `setThemes` to `decoratorThemesReceived`

Commit:
(storyspecific-themes) Rename `selectTheme` to `panelThemeSelected`

Commit:
(storyspecific-themes) Rename `theme` to `stateTheme`

Commit:
(storyspecific-themes) Rename `setTheme` to `setStateTheme`

Commit:
(storyspecific-themes) Rename `themes` to `stateThemes`

Commit:
(storyspecific-themes) Rename `setThemes` to `setStateThemes`

Commit:
(storyspecific-themes) panelThemeSelected: pass theme instead of name

Commit:
(storyspecific-themes) Create src/constants.tsx

Commit:
(storyspecific-themes) Implement storyThemesReceived functionality

Commit:
(storyspecific-themes) withThemesProvider: default value for `themes`

Commit:
(storyspecific-themes) Refactor fontFamily to avoid TSLint issue

Commit:
(storyspecific-themes) Improve React keys for theme buttons

See https://reactjs.org/docs/lists-and-keys.html#keys

"The best way to pick a key is to use a string that uniquely identifies
a list item among its siblings. Most often you would use IDs from your
data as keys. When you don't have stable IDs for rendered items, you
may use the item index as a key as a last resort."

See https://github.com/yannickcr/eslint-plugin-react/blob/v7.14.3/docs/rules/no-array-index-key.md

"It's a bad idea to use the array index since it doesn't uniquely
identify your elements."

Commit:
(storyspecific-themes) Implement singleThemeMessage functionality

Commit:
(storyspecific-themes) Implement showSingleThemeButton functionality

Commit:
(storyspecific-themes) Implement buttonAttributes functionality

Commit:
(storyspecific-themes) Update Themes.spec.tsx

Commit:
(storyspecific-themes) Update README.md

Commit:
(storyspecific-themes) onReceiveThemes(): need to call setQueryParams()

stevebutler2210 commented on README.md:
````diff
@@ -45,11 +50,60 @@ storiesOf("demo", module)
+```
+
+
+## Remind
````
If we're updating a bunch of this anyway, may make sense to update this to say `Reminder`? In the context of the content below it the way it's written at the moment reads a little strangely.

kyanmichael replied:
Agreed. Updated in e4a1df9 :thumbsup:

stevebutler2210 commented on src/Themes.tsx:
```diff
@@ -1,75 +1,160 @@
+            )}
+            {(stateThemes.size > 1 || showSingleThemeButton) && (
+                <div style={RowStyle}>
+                    {stateThemes.map((th: Theme) => {
```
Can we assign the Theme argument here to `theme` instead? It looks like the main reason this wasn't done previously was because it would've clashed, but that is no longer a problem. Not a major issue, but `th` isn't particularly descriptive (also think of `<th>` elements when I see `th` in code, which could be a minor point of confusion)

kyanmichael replied:
Renamed in e4a1df9 :thumbsup:

stevebutler2210 commented on src/Themes.tsx:
```diff
@@ -1,75 +1,160 @@
+            {stateThemes.size === 1 && singleThemeMessage && (
+                <div style={MessageStyle}>{singleThemeMessage}</div>
+            )}
+            {(stateThemes.size > 1 || showSingleThemeButton) && (
```
Might be nice to break some of this down into smaller functions.

E.g.
```
const showThemeButtons = (themes, force = false) => themes.size > 1 || force;
```

Then you could have 2 separate functions which return the content for either the singleThemeMessage or the theme buttons, and have this become:

```
<div>
  {showThemeButtons(stateThemes, showSingleThemeButton) ? renderMessage() : renderButtons()}
</div>
```

stevebutler2210 replied:
As discussed offline, happy to waive this now 👍

kyanmichael replied:
Cool, thanks :thumbsup:

stevebutler2210 commented on src/Themes.tsx:
```diff
@@ -1,75 +1,160 @@
+                    {stateThemes.map((th: Theme) => {
+                        const style: React.CSSProperties = ((th === stateTheme) ? SelectedButtonStyle : ButtonStyle);
+                        const themeNameProps: object = buttonAttributes.reduce(
+                            (obj: object, attr: string) => ({ ...obj, [attr]: th.name}), {},
```
Is this just going to always set the value of every buttonAttribute passed to `th.name`? If so that doesn't feel particularly useful - maybe `buttonAttributes` would be better as a `Map`? (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)

stevebutler2210 replied:
@kyanmichael as discussed internally in Slack, the buttonAttributes are now surplus to requirements and can be removed 😄

kyanmichael replied:
Indeed! Many thanks for the discussion :slightly_smiling_face: Removed in e0af2cd :thumbsup:

stevebutler2210 commented on src/Themes.tsx:
```diff
@@ -1,75 +1,160 @@
+                        );
+                        return (
+                            <div style={style} key={th.name} onClick={() => onSelectTheme(th)} {...themeNameProps}>
+                                {th.name}
```
We're calling `th.name` quite a bit here - might be worth assigning it to a variable?

kyanmichael replied:
After removing the `buttonAttributes` functionality, `th.name` (now renamed to `theme.name`) is only used in 2 places, so it's probably not worth assigning to a variable.

stevebutler2210 reviewed (commented):
Couple of thoughts - nothing major 👍

Commit:
(storyspecific-themes) Minor cleanup in README.md and src/Themes.tsx

stevebutler2210 reviewed (changes requested):
Should be good to go once the buttonAttributes are done with 👍

Commit:
(storyspecific-themes) Remove buttonAttributes functionality

After further thought, there's probably no
realistic use-case for this functionality.

stevebutler2210 reviewed (approved):
Looks good to me!
